### PR TITLE
Fix handling of "for x in x" homonyms

### LINKED ISSUE #6136: False positive `unused-variable` for homonyms between comprehension variable and outer container
### Bug description

```python
def f():
    project_id = []  # this ought to be called project_ids, but...
    assert (project_id > 0 for project_id in project_id)
```

### Configuration

_No response_

### Command used

```shell
pylint z.py
```

### Pylint output

```shell
************* Module z
z.py:2:4: W0612: Unused variable 'project_id' (unused-variable)

-------------------------------------------------------------------
Your code has been rated at 6.67/10 (previous run: 10.00/10, -3.33)
```

### Expected behavior

No msg

### Pylint version

Regression in #6073. Found it by diffing primer package messages.

### OS / Environment

_No response_

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Closes #6136

Extends the fix from #6073. (There, I had to be careful to not emit `unused-argument`, since that's all the test suite alerted me to, but with diffing primer output locally, I noticed the same deal with `unused-variable`. So I handle that the same way.)

### CODE REVIEW

File: tests/functional/u/undefined/undefined_variable_py38.py [hunk 1]
```diff
@@ -123,7 +123,7 @@ def type_annotation_used_after_comprehension():
 
 def type_annotation_unused_after_comprehension():
     """https://github.com/PyCQA/pylint/issues/5326"""
-    my_int: int  # [unused-variable]
+    my_int: int
```
Comment (jacobtylerwalls) [inline tests/functional/u/undefined/undefined_variable_py38.py:126]: This message was added in #5711 without much conviction. It probably shouldn't be here at all. Removing this is a better fix for the issue in #5326, which is a ticket I opened to say _hey we started emitting a message here!_

File: ChangeLog [hunk 2]
```diff
@@ -66,6 +66,11 @@ Release date: TBA
 
   Closes #6069
 
+* Fix false positive for ``unused-variable`` (in unreleased development) for
+  patterns like ``[_ for same_name in same_name]``.
+
+  Closes #6136
+
```
Comment (Pierre-Sassoulas) [inline ChangeLog:73]:
```suggestion
  Closes #6136
```
I suppose we can reduce the changelog if it's all fixed before we release.

**Review by Pierre-Sassoulas — APPROVED** [review]
👍

File: ChangeLog [hunk 3]
```diff
@@ -65,7 +65,7 @@ Release date: TBA
   subscripts in comprehensions.
 
   Closes #6069
-
+  Closes #6136
```
Comment (Pierre-Sassoulas) [inline ChangeLog:68]:
```suggestion
  Closes #6136

```

**Pull request merged**